### PR TITLE
Speed up rebuilding of the user directory for local users

### PR DESCRIPTION
**Pull request by erikjohnston**
The idea here is to batch up the work.

Reviewable commit-by-commit.

Note the second commit is simply moving a function out, and so reviewing with whitespace changes disabled is recommended.

### CODE REVIEW

File: synapse/storage/databases/main/user_directory.py [hunk 1]
```diff
@@ -596,75 +609,99 @@ async def update_profile_in_user_dir(
...
             table="user_directory",
-            keyvalues={"user_id": user_id},
-            values={"display_name": display_name, "avatar_url": avatar_url},
+            key_names=("user_id",),
+            key_values=[(p.user_id,) for p in profiles],
+            value_names=("display_name", "avatar_url"),
+            value_values=[
+                (
+                    non_null_str_or_none(p.display_name),
+                    non_null_str_or_none(p.avatar_url),
```
Comment (clokep) [inline synapse/storage/databases/main/user_directory.py:664]: I don't really see the equivalent of these somewhere in the single-profile version?

File: synapse/storage/databases/main/user_directory.py [hunk 2]
```diff
@@ -596,75 +609,99 @@ async def update_profile_in_user_dir(
         await self.db_pool.runInteraction(
             "update_profile_in_user_dir",
             self._update_profile_in_user_dir_txn,
-            user_id,
-            display_name,
-            avatar_url,
+            [_UserDirProfile(user_id, display_name, avatar_url)],
         )
 
     def _update_profile_in_user_dir_txn(
```
Comment (clokep) [inline synapse/storage/databases/main/user_directory.py:650]: Should we rename to `update_profiles*s*`?

**Review by clokep — APPROVED** [review]
Seems reasonable to me.

**Comment by erikjohnston** [comment]
c.f. #15264

**Pull request merged**